# Nashville-Housing

## Introduction
Real world datasets are not always clean to work with. They consists of missing values, duplicate values and other noticeable differences. With the help of data cleaning we can process and prepare the dataset for analysis. Data analysts and data scientists spends more than 70% of their time in cleaning and perfecting the datasets to work with. 

A very common scenario is that the dataset contains missing values coded as NaN. Also, the missing values are coded in different ways. The dataset may contain negative or invalid values. It may contain outliers. It may be in the untidy format. All of these are examples of a messy dataset

In [75]:
!pip install pandas --upgrade --quiet
!pip install jovian --upgrade --quiet
!pip install numpy --quiet

We need two Python libraries for the data cleansing process – NumPy and Pandas.

• NumPy – NumPy is the fundamental Python library for scientific computing. It adds support for large and multi-dimensional arrays and matrices. It also supports large collection of high-level mathematical functions to operate on these arrays.

• Pandas - Pandas is a software library for Python programming language which provide tools for data manipulation and analysis tasks. It will enable us to manipulate numerical tables and time series using data structures and operations.

## Import the required Python libraries
We need two Python libraries – NumPy and Pandas for the data cleaning process. We need to import these libraries before we actually start using them. We can import them with their usual shorthand notation as follows:-

In [76]:
import jovian
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
pd.options.display.float_format = '{:.2f}'.format #for showing the integer data set with 2 decimal points

## Sourcing the dataset
For this project, I have downloaded a dataset from kaggle. The dataset consists of details of Nashville housing market data.

In [77]:
data_df = pd.read_csv("Nashville_housing_data_2013_2016.csv")
data_df

Unnamed: 0.1  Unnamed: 0         Parcel ID           Land Use  \
0                 0           0  105 03 0D 008.00  RESIDENTIAL CONDO   
1                 1           1   105 11 0 080.00      SINGLE FAMILY   
2                 2           2   118 03 0 130.00      SINGLE FAMILY   
3                 3           3   119 01 0 479.00      SINGLE FAMILY   
4                 4           4   119 05 0 186.00      SINGLE FAMILY   
...             ...         ...               ...                ...   
56631         56631       56631  093 13 0B 274.00  RESIDENTIAL CONDO   
56632         56632       56632  093 13 0D 044.00  RESIDENTIAL CONDO   
56633         56633       56633  093 13 0D 048.00  RESIDENTIAL CONDO   
56634         56634       56634  093 13 0D 056.00  RESIDENTIAL CONDO   
56635         56635       56635  093 13 0D 094.00  RESIDENTIAL CONDO   

        Property Address Suite/ Condo   # Property City   Sale Date  \
0        1208  3RD AVE S                8     NASHVILLE  2013-01-24   
1       1802  STEWART PL              NaN     NASHVILLE  2013-01-11   
2      2761  ROSEDALE PL              NaN     NASHVILLE  2013-01-18   
3      224  PEACHTREE ST              NaN     NASHVILLE  2013-01-18   
4          316  LUTIE ST              NaN     NASHVILLE  2013-01-23   
...                  ...              ...           ...         ...   
56631    320 11TH  AVE S            274.0     NASHVILLE  2016-10-06   
56632    700 12TH  AVE S            608.0     NASHVILLE  2016-10-25   
56633    700 12TH  AVE S            613.0     NASHVILLE  2016-10-04   
56634    700 12TH  AVE S            708.0     NASHVILLE  2016-10-26   
56635    700 12TH  AVE S           1008.0     NASHVILLE  2016-10-27   

       Sale Price   Legal Reference Sold As Vacant  \
0          132000  20130128-0008725             No   
1          191500  20130118-0006337             No   
2          202000  20130124-0008033             No   
3           32000  20130128-0008863             No   
4          102000  20130131-0009929             No   
...           ...               ...            ...   
56631      210000  20161007-0106599             No   
56632      338000  20161101-0115186             No   
56633      742000  20161010-0106889             No   
56634      320000  20161031-0114730             No   
56635      330000  20161104-0117077             No   

      Multiple Parcels Involved in Sale                     Owner Name  \
0                                    No                            NaN   
1                                    No              STINSON, LAURA M.   
2                                    No                NUNES, JARED R.   
3                                    No                WHITFORD, KAREN   
4                                    No  HENDERSON, JAMES P. & LYNN P.   
...                                 ...                            ...   
56631                                No                            NaN   
56632                                No                            NaN   
56633                                No                            NaN   
56634                                No                            NaN   
56635                                No                            NaN   

                 Address       City State  Acreage             Tax District  \
0                    NaN        NaN   NaN      NaN                      NaN   
1       1802  STEWART PL  NASHVILLE    TN     0.17  URBAN SERVICES DISTRICT   
2      2761  ROSEDALE PL  NASHVILLE    TN     0.11       CITY OF BERRY HILL   
3      224  PEACHTREE ST  NASHVILLE    TN     0.17  URBAN SERVICES DISTRICT   
4          316  LUTIE ST  NASHVILLE    TN     0.34  URBAN SERVICES DISTRICT   
...                  ...        ...   ...      ...                      ...   
56631                NaN        NaN   NaN      NaN                      NaN   
56632                NaN        NaN   NaN      NaN                      NaN   
56633                NaN        NaN   Na

### Insights:

- According to the information about the data in Kaggle site, this dataset is missing about half of the home details data. 

#### For a safer side, we will copy the dataset and make changes to the new copied version of it. So that if we need the original dataset later, we can access it.

In [78]:
#copying the dataset 

housing_df = data_df.copy(deep= True)

In [79]:
housing_df

Unnamed: 0.1  Unnamed: 0         Parcel ID           Land Use  \
0                 0           0  105 03 0D 008.00  RESIDENTIAL CONDO   
1                 1           1   105 11 0 080.00      SINGLE FAMILY   
2                 2           2   118 03 0 130.00      SINGLE FAMILY   
3                 3           3   119 01 0 479.00      SINGLE FAMILY   
4                 4           4   119 05 0 186.00      SINGLE FAMILY   
...             ...         ...               ...                ...   
56631         56631       56631  093 13 0B 274.00  RESIDENTIAL CONDO   
56632         56632       56632  093 13 0D 044.00  RESIDENTIAL CONDO   
56633         56633       56633  093 13 0D 048.00  RESIDENTIAL CONDO   
56634         56634       56634  093 13 0D 056.00  RESIDENTIAL CONDO   
56635         56635       56635  093 13 0D 094.00  RESIDENTIAL CONDO   

        Property Address Suite/ Condo   # Property City   Sale Date  \
0        1208  3RD AVE S                8     NASHVILLE  2013-01-24   
1       1802  STEWART PL              NaN     NASHVILLE  2013-01-11   
2      2761  ROSEDALE PL              NaN     NASHVILLE  2013-01-18   
3      224  PEACHTREE ST              NaN     NASHVILLE  2013-01-18   
4          316  LUTIE ST              NaN     NASHVILLE  2013-01-23   
...                  ...              ...           ...         ...   
56631    320 11TH  AVE S            274.0     NASHVILLE  2016-10-06   
56632    700 12TH  AVE S            608.0     NASHVILLE  2016-10-25   
56633    700 12TH  AVE S            613.0     NASHVILLE  2016-10-04   
56634    700 12TH  AVE S            708.0     NASHVILLE  2016-10-26   
56635    700 12TH  AVE S           1008.0     NASHVILLE  2016-10-27   

       Sale Price   Legal Reference Sold As Vacant  \
0          132000  20130128-0008725             No   
1          191500  20130118-0006337             No   
2          202000  20130124-0008033             No   
3           32000  20130128-0008863             No   
4          102000  20130131-0009929             No   
...           ...               ...            ...   
56631      210000  20161007-0106599             No   
56632      338000  20161101-0115186             No   
56633      742000  20161010-0106889             No   
56634      320000  20161031-0114730             No   
56635      330000  20161104-0117077             No   

      Multiple Parcels Involved in Sale                     Owner Name  \
0                                    No                            NaN   
1                                    No              STINSON, LAURA M.   
2                                    No                NUNES, JARED R.   
3                                    No                WHITFORD, KAREN   
4                                    No  HENDERSON, JAMES P. & LYNN P.   
...                                 ...                            ...   
56631                                No                            NaN   
56632                                No                            NaN   
56633                                No                            NaN   
56634                                No                            NaN   
56635                                No                            NaN   

                 Address       City State  Acreage             Tax District  \
0                    NaN        NaN   NaN      NaN                      NaN   
1       1802  STEWART PL  NASHVILLE    TN     0.17  URBAN SERVICES DISTRICT   
2      2761  ROSEDALE PL  NASHVILLE    TN     0.11       CITY OF BERRY HILL   
3      224  PEACHTREE ST  NASHVILLE    TN     0.17  URBAN SERVICES DISTRICT   
4          316  LUTIE ST  NASHVILLE    TN     0.34  URBAN SERVICES DISTRICT   
...                  ...        ...   ...      ...                      ...   
56631                NaN        NaN   NaN      NaN                      NaN   
56632                NaN        NaN   NaN      NaN                      NaN   
56633                NaN        NaN   Na

## Exploratory Data Analysis
Now, it is time to understand the data. We should diagnose the data for any discrepancies by doing exploratory data analysis. We should proceed as follows:-

df.shape attribute
We can check the dimensions of the data with df.shape attribute.


In [80]:
housing_df.shape

(56636, 31)

### Insights
- Our dataset has total of 31 columns and 56636 rows.
- 2 of the columns are identical which we will remove in our later task.

In [81]:
# checking for null values
housing_df.isnull().any()

Unnamed: 0.1                         False
Unnamed: 0                           False
Parcel ID                            False
Land Use                             False
Property Address                      True
Suite/ Condo   #                      True
Property City                         True
Sale Date                            False
Sale Price                           False
Legal Reference                      False
Sold As Vacant                       False
Multiple Parcels Involved in Sale    False
Owner Name                            True
Address                               True
City                                  True
State                                 True
Acreage                               True
Tax District                          True
Neighborhood                          True
image                                 True
Land Value                            True
Building Value                        True
Total Value                           True
Finished Ar

## df.info() method
We can get a concise summary of the dataset with df.info() method. This method prints information about a DataFrame including the index dtype and column dtypes, non-null values and memory usage.

In [82]:
housing_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56636 entries, 0 to 56635
Data columns (total 31 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Unnamed: 0.1                       56636 non-null  int64  
 1   Unnamed: 0                         56636 non-null  int64  
 2   Parcel ID                          56636 non-null  object 
 3   Land Use                           56636 non-null  object 
 4   Property Address                   56477 non-null  object 
 5   Suite/ Condo   #                   6109 non-null   object 
 6   Property City                      56477 non-null  object 
 7   Sale Date                          56636 non-null  object 
 8   Sale Price                         56636 non-null  int64  
 9   Legal Reference                    56636 non-null  object 
 10  Sold As Vacant                     56636 non-null  object 
 11  Multiple Parcels Involved in Sale  56636 non-null  obj

### df.head() and df.tail() methods
We can view the top five and bottom five rows of the dataset with df.head() and df.tail() methods respectively.

In [89]:
housing_df.head()

Unnamed: 0.1  Unnamed: 0         Parcel ID           Land Use  \
0             0           0  105 03 0D 008.00  RESIDENTIAL CONDO   
1             1           1   105 11 0 080.00      SINGLE FAMILY   
2             2           2   118 03 0 130.00      SINGLE FAMILY   
3             3           3   119 01 0 479.00      SINGLE FAMILY   
4             4           4   119 05 0 186.00      SINGLE FAMILY   

    Property Address Suite/ Condo   # Property City  Sale Date  Sale Price  \
0    1208  3RD AVE S                8     NASHVILLE 2013-01-24      132000   
1   1802  STEWART PL              NaN     NASHVILLE 2013-01-11      191500   
2  2761  ROSEDALE PL              NaN     NASHVILLE 2013-01-18      202000   
3  224  PEACHTREE ST              NaN     NASHVILLE 2013-01-18       32000   
4      316  LUTIE ST              NaN     NASHVILLE 2013-01-23      102000   

    Legal Reference Sold As Vacant Multiple Parcels Involved in Sale  \
0  20130128-0008725             No                                No   
1  20130118-0006337             No                                No   
2  20130124-0008033             No                                No   
3  20130128-0008863             No                                No   
4  20130131-0009929             No                                No   

                      Owner Name            Address       City State  Acreage  \
0                            NaN                NaN        NaN   NaN      NaN   
1              STINSON, LAURA M.   1802  STEWART PL  NASHVILLE    TN     0.17   
2                NUNES, JARED R.  2761  ROSEDALE PL  NASHVILLE    TN     0.11   
3                WHITFORD, KAREN  224  PEACHTREE ST  NASHVILLE    TN     0.17   
4  HENDERSON, JAMES P. & LYNN P.      316  LUTIE ST  NASHVILLE    TN     0.34   

              Tax District  Neighborhood               image  Land Value  \
0                      NaN           NaN                 NaN         NaN   
1  URBAN SERVICES DISTRICT       3127.00  \114000\910001.JPG    32000.00   
2       CITY OF BERRY HILL       9126.00  \131000\191001.JPG    34000.00   
3  URBAN SERVICES DISTRICT       3130.00  \133000\721001.JPG    25000.00   
4  URBAN SERVICES DISTRICT       3130.00  \134000\474001.JPG    25000.00   

   Building Value  Total Value  Finished Area Foundation Type  Year Built  \
0             NaN          NaN            NaN             NaN         NaN   
1       134400.00    168300.00        1149.00         PT BSMT     1941.00   
2       157800.00    191800.00        2090.82            SLAB     2000.00   
3       243700.00    268700.00        2145.60       FULL BSMT     1948.00   
4       138100.00    164800.00        1969.00           CRAWL     1910.00   

  Exterior Wall Grade  Bedrooms  Full Bath  Half Bath  
0           NaN   NaN       NaN        NaN        NaN  
1         BRICK  C         2.00       1.00       0.00  
2   BRICK/FRAME  C         3.00       2.00       1.00  
3   BRICK/FRAME  B         4.00       2.00       0.00  
4         FRAME  C         2.00       1.00       0.00

In [88]:
housing_df.tail()

Unnamed: 0.1  Unnamed: 0         Parcel ID           Land Use  \
56631         56631       56631  093 13 0B 274.00  RESIDENTIAL CONDO   
56632         56632       56632  093 13 0D 044.00  RESIDENTIAL CONDO   
56633         56633       56633  093 13 0D 048.00  RESIDENTIAL CONDO   
56634         56634       56634  093 13 0D 056.00  RESIDENTIAL CONDO   
56635         56635       56635  093 13 0D 094.00  RESIDENTIAL CONDO   

      Property Address Suite/ Condo   # Property City  Sale Date  Sale Price  \
56631  320 11TH  AVE S            274.0     NASHVILLE 2016-10-06      210000   
56632  700 12TH  AVE S            608.0     NASHVILLE 2016-10-25      338000   
56633  700 12TH  AVE S            613.0     NASHVILLE 2016-10-04      742000   
56634  700 12TH  AVE S            708.0     NASHVILLE 2016-10-26      320000   
56635  700 12TH  AVE S           1008.0     NASHVILLE 2016-10-27      330000   

        Legal Reference Sold As Vacant Multiple Parcels Involved in Sale  \
56631  20161007-0106599             No                                No   
56632  20161101-0115186             No                                No   
56633  20161010-0106889             No                                No   
56634  20161031-0114730             No                                No   
56635  20161104-0117077             No                                No   

      Owner Name Address City State  Acreage Tax District  Neighborhood image  \
56631        NaN     NaN  NaN   NaN      NaN          NaN           NaN   NaN   
56632        NaN     NaN  NaN   NaN      NaN          NaN           NaN   NaN   
56633        NaN     NaN  NaN   NaN      NaN          NaN           NaN   NaN   
56634        NaN     NaN  NaN   NaN      NaN          NaN           NaN   NaN   
56635        NaN     NaN  NaN   NaN      NaN          NaN           NaN   NaN   

       Land Value  Building Value  Total Value  Finished Area Foundation Type  \
56631         NaN             NaN          NaN            NaN             NaN   
56632         NaN             NaN          NaN            NaN             NaN   
56633         NaN             NaN          NaN            NaN             NaN   
56634         NaN             NaN          NaN            NaN             NaN   
56635         NaN             NaN          NaN            NaN             NaN   

       Year Built Exterior Wall Grade  Bedrooms  Full Bath  Half Bath  
56631         NaN           NaN   NaN       NaN        NaN        NaN  
56632         NaN           NaN   NaN       NaN        NaN        NaN  
56633         NaN           NaN   NaN       NaN        NaN        NaN  
56634         NaN           NaN   NaN       NaN        NaN        NaN  
56635         NaN           NaN   NaN       NaN        NaN        NaN

### Insights:
As we can see that there are a lot of discrepancies in our data set.
- The property address is not defined properly.
- Some of the columns hasn't been formatted correctly.
- There are so many columns with personal information such as owner's name, legal referance number etc. 

### df.dtypes attribute
We can check the data types of each column in the dataframe with df.dtypes attribute. The above command returns the data type of each column

In [90]:
housing_df.dtypes

Unnamed: 0.1                                  int64
Unnamed: 0                                    int64
Parcel ID                                    object
Land Use                                     object
Property Address                             object
Suite/ Condo   #                             object
Property City                                object
Sale Date                            datetime64[ns]
Sale Price                                    int64
Legal Reference                              object
Sold As Vacant                               object
Multiple Parcels Involved in Sale            object
Owner Name                                   object
Address                                      object
City                                         object
State                                        object
Acreage                                     float64
Tax District                                 object
Neighborhood                                float64
image       

### Insights:
- We can see that the data type for 'Sale Date' column is not correct. The sale date data type must be of datetime format.

### 1. Standardize Date Format
As we can see in the dataset info, the "Sale Date" column is set to an object type data which is not accurate given the information about the data. in the first question of this project, we will convert the data type of this column from object to 'datetime' format.

In [91]:
housing_df['Sale Date'] = pd.to_datetime(housing_df['Sale Date'])
housing_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56636 entries, 0 to 56635
Data columns (total 31 columns):
 #   Column                             Non-Null Count  Dtype         
---  ------                             --------------  -----         
 0   Unnamed: 0.1                       56636 non-null  int64         
 1   Unnamed: 0                         56636 non-null  int64         
 2   Parcel ID                          56636 non-null  object        
 3   Land Use                           56636 non-null  object        
 4   Property Address                   56477 non-null  object        
 5   Suite/ Condo   #                   6109 non-null   object        
 6   Property City                      56477 non-null  object        
 7   Sale Date                          56636 non-null  datetime64[ns]
 8   Sale Price                         56636 non-null  int64         
 9   Legal Reference                    56636 non-null  object        
 10  Sold As Vacant                    

### df.describe() method
We can view the summary statistics of numerical columns with df.describe() method. It enable us to detect outliers in the data which require further investigation.

In [92]:
housing_df.describe()

Unnamed: 0.1  Unnamed: 0                      Sale Date  Sale Price  \
count      56636.00    56636.00                          56636    56636.00   
mean       28317.50    28317.50  2015-02-07 21:40:32.459919616   327211.13   
min            0.00        0.00            2013-01-02 00:00:00       50.00   
25%        14158.75    14158.75            2014-04-11 00:00:00   135000.00   
50%        28317.50    28317.50            2015-03-24 00:00:00   205450.00   
75%        42476.25    42476.25            2015-12-30 00:00:00   329000.00   
max        56635.00    56635.00            2016-10-31 00:00:00 54278060.00   
std        16349.55    16349.55                            NaN   928742.55   

       Acreage  Neighborhood  Land Value  Building Value  Total Value  \
count 26017.00      26017.00    26017.00        26017.00     26017.00   
mean      0.50       4356.22    69072.67       160802.46    232397.09   
min       0.01        107.00      100.00            0.00       100.00   
25%       0.18       3126.00    21000.00        75900.00    102800.00   
50%       0.27       3929.00    28800.00       111400.00    148500.00   
75%       0.45       6228.00    60000.00       180700.00    268500.00   
max     160.06       9530.00  2772000.00     12971800.00  13940400.00   
std       1.57       2170.35   106040.53       206804.06    281070.27   

       Finished Area  Year Built  Bedrooms  Full Bath  Half Bath  
count       24166.00    24165.00  24159.00   24277.00   24146.00  
mean         1926.95     1963.75      3.09       1.89       0.28  
min             0.00     1799.00      0.00       0.00       0.00  
25%          1239.00     1948.00      3.00       1.00       0.00  
50%          1632.00     1960.00      3.00       2.00       0.00  
75%          2212.00     1983.00      3.00       2.00       1.00  
max        197988.00     2017.00     11.00      10.00       3.00  
std          1687.02       26.55      0.85       0.96       0.49

### Insights:
- Thankfully in our dataset there are no discrepancies in related to statistical terms. 
- The mean, min and max values are all positive and not null.
- By performing the summery statistics we can say that the dataset does not contains any outliers.

### 2. Populate Property Address data
Out of total 56636 data entries in the data set, the 'Property Address' column has total of 56477 "non-null" values. This implies that total of 159 rows has null values in the column (some of them are missing). We will perform this task in a step by step manner from finding the duplicate values to joining them in the necessary place

- Some of the 'Parcel IDs' have been entered multiple times in the dataset with some of them having 'Property Address' and some of them with null property address.

In [93]:
# checking for duplicate parcel ids
duplicate_parcel = housing_df.duplicated(subset= 'Parcel ID', keep=False)
duplicate_parcel

0        False
1        False
2        False
3         True
4         True
         ...  
56631     True
56632    False
56633     True
56634     True
56635    False
Length: 56636, dtype: bool

- After checking the duplicate parcel ids we will extract the one which has null values

In [94]:
# extracting the parcel id with null property addresses
null_address = housing_df.loc[housing_df['Property Address'].isnull(), 'Parcel ID']
null_address

1124     132 05 0 119.00
1713     104 01 0 374.00
1767     105 05 0 556.00
1820     117 07 0 147.00
1842     118 01 0 265.00
              ...       
41710    131 06 0 022.00
42068    072 11 0 177.00
43215    072 11 0 402.00
43848    116 04 0 118.00
44211    072 15 0 176.00
Name: Parcel ID, Length: 159, dtype: object

- In this step, we will find the duplicate values to replace with the missing one. 
  We will create a new dataframe with rows where the parcel id has no duplicate values and the property address is not null. 
  We will use the '&' operator to combine the above two conditions.

In [95]:
valid_add = housing_df.loc[~duplicate_parcel & ~housing_df['Property Address'].isnull(), ['Parcel ID', 'Property Address']]   #using ~ operator to represent not
valid_add

Parcel ID    Property Address
0      105 03 0D 008.00     1208  3RD AVE S
1       105 11 0 080.00    1802  STEWART PL
2       118 03 0 130.00   2761  ROSEDALE PL
6      119 10 0A 104.00    104  PRESCOTT PL
10      133 07 0 195.00    184  WHEELER AVE
...                 ...                 ...
56624  082 14 0A 144.00      732 1ST  AVE N
56626  093 06 0A 004.00  301 DEMONBREUN  ST
56630  093 11 0B 012.00      464 2ND  AVE S
56632  093 13 0D 044.00     700 12TH  AVE S
56635  093 13 0D 094.00     700 12TH  AVE S

[41319 rows x 2 columns]

- To find the valid addresses based on the Parcel ID We'll set the Parcel ID as the index of the valid_add DataFrame. This way, we can easily find the Property Address for each Parcel ID.

In [96]:
add_map = valid_add.set_index('Parcel ID')['Property Address']
add_map

Parcel ID
105 03 0D 008.00       1208  3RD AVE S
105 11 0 080.00       1802  STEWART PL
118 03 0 130.00      2761  ROSEDALE PL
119 10 0A 104.00      104  PRESCOTT PL
133 07 0 195.00       184  WHEELER AVE
                           ...        
082 14 0A 144.00        732 1ST  AVE N
093 06 0A 004.00    301 DEMONBREUN  ST
093 11 0B 012.00        464 2ND  AVE S
093 13 0D 044.00       700 12TH  AVE S
093 13 0D 094.00       700 12TH  AVE S
Name: Property Address, Length: 41319, dtype: object

- In the end, we will update the null values using add_book and map function

In [97]:
housing_df.loc[housing_df['Parcel ID'].isin(null_address), 'Property Address'] = housing_df['Parcel ID'].map(add_map)
housing_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56636 entries, 0 to 56635
Data columns (total 31 columns):
 #   Column                             Non-Null Count  Dtype         
---  ------                             --------------  -----         
 0   Unnamed: 0.1                       56636 non-null  int64         
 1   Unnamed: 0                         56636 non-null  int64         
 2   Parcel ID                          56636 non-null  object        
 3   Land Use                           56636 non-null  object        
 4   Property Address                   56462 non-null  object        
 5   Suite/ Condo   #                   6109 non-null   object        
 6   Property City                      56477 non-null  object        
 7   Sale Date                          56636 non-null  datetime64[ns]
 8   Sale Price                         56636 non-null  int64         
 9   Legal Reference                    56636 non-null  object        
 10  Sold As Vacant                    

### 3. Breaking out Address into Individual Columns (Address, City, State)

In [98]:
print(housing_df['Property Address'].head())

0      1208  3RD AVE S
1     1802  STEWART PL
2    2761  ROSEDALE PL
3    224  PEACHTREE ST
4        316  LUTIE ST
Name: Property Address, dtype: object


In [99]:
# Splitting Property Address in to seperate columns

housing_df[['PAddress', 'PCity', 'PState']] = housing_df['Property Address'].str.split(r'\s+', expand=True, n=2)
housing_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56636 entries, 0 to 56635
Data columns (total 34 columns):
 #   Column                             Non-Null Count  Dtype         
---  ------                             --------------  -----         
 0   Unnamed: 0.1                       56636 non-null  int64         
 1   Unnamed: 0                         56636 non-null  int64         
 2   Parcel ID                          56636 non-null  object        
 3   Land Use                           56636 non-null  object        
 4   Property Address                   56462 non-null  object        
 5   Suite/ Condo   #                   6109 non-null   object        
 6   Property City                      56477 non-null  object        
 7   Sale Date                          56636 non-null  datetime64[ns]
 8   Sale Price                         56636 non-null  int64         
 9   Legal Reference                    56636 non-null  object        
 10  Sold As Vacant                    

### 4. a. Dealing with missing numerical values
The following commands help us to deal with missing numerical values.

df.isnull() :
This command checks whether each cell in a dataframe contains missing values or not. If the cell contains missing value, it returns True otherwise it returns False.

df.isnull.sum() :
This command returns the total number of missing values in each column in the dataset.

In [100]:
housing_df.isnull().sum()

Unnamed: 0.1                             0
Unnamed: 0                               0
Parcel ID                                0
Land Use                                 0
Property Address                       174
Suite/ Condo   #                     50527
Property City                          159
Sale Date                                0
Sale Price                               0
Legal Reference                          0
Sold As Vacant                           0
Multiple Parcels Involved in Sale        0
Owner Name                           31375
Address                              30619
City                                 30619
State                                30619
Acreage                              30619
Tax District                         30619
Neighborhood                         30619
image                                31301
Land Value                           30619
Building Value                       30619
Total Value                          30619
Finished Ar

### 4. b. Fill missing values with a test statistic
We will fill the missing numeric values of the column 'Land Value', 'Building Value', 'Total Value' with a test statistic like mean, median or mode of the particular feature the missing value belongs to.

Filling missing values with a test statistic like mean
mean = df['col_name'].mean()

df['col_name'].fillna(value = median, inplace = True )

In [101]:
mean_land_value = housing_df['Land Value'].mean()

housing_df['Land Value'].fillna(mean_land_value, inplace=True)

In [102]:
mean_building_value = housing_df['Building Value'].mean()

housing_df['Building Value'].fillna(mean_building_value, inplace=True)

In [103]:
mean_total_value = housing_df['Total Value'].mean()
housing_df['Total Value'].fillna(mean_total_value, inplace=True)

In [104]:
housing_df.isnull().sum()

Unnamed: 0.1                             0
Unnamed: 0                               0
Parcel ID                                0
Land Use                                 0
Property Address                       174
Suite/ Condo   #                     50527
Property City                          159
Sale Date                                0
Sale Price                               0
Legal Reference                          0
Sold As Vacant                           0
Multiple Parcels Involved in Sale        0
Owner Name                           31375
Address                              30619
City                                 30619
State                                30619
Acreage                              30619
Tax District                         30619
Neighborhood                         30619
image                                31301
Land Value                               0
Building Value                           0
Total Value                              0
Finished Ar

### 5. Delete Unused Columns
in our last task we will drop the columns which doesnot provide any insights and we might not need during analysis.

In [105]:
#dropping the unused columns
unused_col = ['Unnamed: 0.1','Unnamed: 0', 'Legal Reference', 'Owner Name', 'Address', 'City', 'State']
housing_df.drop( columns= unused_col, axis=1, inplace= True)
housing_df.columns

Index(['Parcel ID', 'Land Use', 'Property Address', 'Suite/ Condo   #',
       'Property City', 'Sale Date', 'Sale Price', 'Sold As Vacant',
       'Multiple Parcels Involved in Sale', 'Acreage', 'Tax District',
       'Neighborhood', 'image', 'Land Value', 'Building Value', 'Total Value',
       'Finished Area', 'Foundation Type', 'Year Built', 'Exterior Wall',
       'Grade', 'Bedrooms', 'Full Bath', 'Half Bath', 'PAddress', 'PCity',
       'PState'],
      dtype='object')

## Future Works

There are lot of scopes of improvement can be added in this project in future, with the data provided we can,

- Conducting a data visualization.
- Determine the percentage of the house sold.
- We can compare the sale proce with the actual property value.
- Create ML algo to predict future house sales based on the provided data.

In [ ]:
import jovian
jovian.commit(project= 'nashville-housing-dc')

<IPython.core.display.Javascript object>